ibm/service/power/resource_ibm_pi_instance.go

```diff
@@ -531,13 +546,19 @@ func resourceIBMPIInstanceUpdate(ctx context.Context, d *schema.ResourceData, me
 	}
 	cores_enabled := checkCloudInstanceCapability(cloudInstance, CUSTOM_VIRTUAL_CORES)
 
-	if d.HasChange(helpers.PIInstanceName) {
-		body := &models.PVMInstanceUpdate{
-			ServerName: name,
+	if d.HasChanges(helpers.PIInstanceName, PIVirtualOpticalDevice) {
+		body := &models.PVMInstanceUpdate{}
+		if d.HasChange(helpers.PIInstanceName) {
+			body.ServerName = name
+		}
+		if d.HasChange(PIVirtualOpticalDevice) {
+			if vod, ok := d.GetOk(PIVirtualOpticalDevice); ok {
```

Comment:
If value is not provided then we need to update the instance to remove the previous set value of `PIVirtualOpticalDevice`.

Reply:
I think we may not need to do this as the value of optical device is either attach or "".

Reply:
With hadChange it makes sense.

Reply:
When GetOK is false we are not setting `body.CloudInitialization.VirtualOpticalDevice`.

Reply:
Updated.